### PR TITLE
Refactor SqlServerFhirModel initialization

### PR DESCRIPTION
**Pull request by rotodd**
## Description

- Updates `SqlServerFhirModel` to:
  - inherit from `IStartable` instead of `SqlServerModelInitialization`
  - start up the Schema Initializer and the Search Parameter Definition Manager
- Puts `SqlServerModelInitialization.GetStringId()` in `SqlServerFhirModel` (`SqlServerModelInitialization` will be removed in [#AB74343](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74343))
- Makes`SearchParameterDefinitionManager.Start()` idempotent

## Related issues
Addresses [#AB74281](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74281).

## Testing
- All tests pass as before

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs [hunk 1]
```diff
@@ -47,94 +49,94 @@ public sealed class SqlServerFhirModel : SqlServerModelInitializer
...
         }
 
-        protected override async Task Initialize()
+        public void Start()
         {
-            var connectionStringBuilder = new SqlConnectionStringBuilder(SqlServerDataStoreConfiguration.ConnectionString);
+            StartAsync().GetAwaiter().GetResult();
+        }
+
+        // TODO: Add this to IStartable interface.
```
Comment (rotodd) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:127]: I created [#AB74342](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74342) to address this.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs [hunk 2]
```diff
@@ -47,94 +49,94 @@ public sealed class SqlServerFhirModel : SqlServerModelInitializer
...
         {
-            ThrowIfNotInitialized();
             return _quantityCodeToId.TryGetValue(code, out quantityCodeId);
         }
 
-        protected override async Task Initialize()
+        public void Start()
         {
-            var connectionStringBuilder = new SqlConnectionStringBuilder(SqlServerDataStoreConfiguration.ConnectionString);
+            StartAsync().GetAwaiter().GetResult();
```
Comment (rotodd) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:124]: Pulled this from [here](https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development).

File: src/Microsoft.Health.Fhir.Core/Features/Definition/ISearchParameterDefinitionManager.cs [hunk 3]
```diff
@@ -13,7 +14,7 @@ namespace Microsoft.Health.Fhir.Core.Features.Definition
     /// <summary>
     /// Provides mechanism to access search parameter definition.
     /// </summary>
-    public interface ISearchParameterDefinitionManager
+    public interface ISearchParameterDefinitionManager : IStartable
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/ISearchParameterDefinitionManager.cs:17]: I don't think the interface ISearchParameterDefinitionManager should build on IStartable, these seem like unrelated concepts

File: src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs [hunk 4]
```diff
@@ -36,16 +36,21 @@ public SearchParameterDefinitionManager(IModelInfoProvider modelInfoProvider)
...
+
+                builder.Build();
 
-            builder.Build();
+                _typeLookup = builder.ResourceTypeDictionary;
+                UrlLookup = builder.UriDictionary;
 
-            _typeLookup = builder.ResourceTypeDictionary;
-            UrlLookup = builder.UriDictionary;
+                List<string> list = UrlLookup.Values.Where(p => p.Type == ValueSets.SearchParamType.Composite).Select(p => string.Join("|", p.Component.Select(c => UrlLookup[c.DefinitionUrl].Type))).Distinct().ToList();
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs:50]: @jackliums do you remember why we have this list that isn't used? I keep looking at it and wondering

File: src/Microsoft.Health.Fhir.Core/Features/Definition/ISearchParameterDefinitionManager.cs (hunk 3 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.Core/Features/Definition/ISearchParameterDefinitionManager.cs]: Let me know what you think of 138de6b. I removed `IStartable` from the interface but exposed the `Start()` method so that it could be called from `SqlServerFhirModel`.

Comment (rotodd) [reply src/Microsoft.Health.Fhir.Core/Features/Definition/ISearchParameterDefinitionManager.cs]: As discussed, I removed `Start()` from `ISearchParameterDefinitionManager` in b2f049a.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 5]
```diff
@@ -77,7 +77,7 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
 
         public async Task<UpsertOutcome> UpsertAsync(ResourceWrapper resource, WeakETag weakETag, bool allowCreate, bool keepHistory, CancellationToken cancellationToken)
         {
-            await _model.EnsureInitialized();
+            await _model.StartAsync();
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:80]: Why should we have to call StartAsync() from the consumer?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs [hunk 6]
```diff
@@ -47,94 +49,98 @@ public sealed class SqlServerFhirModel : SqlServerModelInitializer
...
         }
 
         public bool TryGetResourceTypeId(string resourceTypeName, out short id)
         {
-            ThrowIfNotInitialized();
             return _resourceTypeToId.TryGetValue(resourceTypeName, out id);
         }
 
         public string GetResourceTypeName(short resourceTypeId)
         {
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:81]: ThrowIfNotInitialized() will fail in a predicatable way with a better error message than something like a null ref

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs [hunk 7]
```diff
@@ -189,7 +195,7 @@ COMMIT TRANSACTION
                     sqlCommand.Parameters.AddWithValue("@claimTypes", commaSeparatedClaimTypes);
                     sqlCommand.Parameters.AddWithValue("@compartmentTypes", commaSeparatedCompartmentTypes);
 
-                    using (SqlDataReader reader = await sqlCommand.ExecuteReaderAsync(CommandBehavior.SequentialAccess))
+                    using (SqlDataReader reader = await sqlCommand.ExecuteReaderAsync(CommandBehavior.SequentialAccess).ConfigureAwait(false))
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:198]: in .net core, this is not needed

Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:198]: Did a bit a of reading about this... is this the reason why?

"_code running in an ASP.NET Core app by default won’t see a custom SynchronizationContext, which lessens the need for ConfigureAwait(false) running in such an environment._" (taken from [here](https://devblogs.microsoft.com/dotnet/configureawait-faq/))

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 6 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:81]: I was under the impression that these methods would never be called before `Start()` and `StartAsync()`, so there wouldn't be a need to check initialization. Please correct me if I'm wrong!

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 7 above)
Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:198]: Yeah. What I said was technically not true, but we're not publishing these as general-purpose libraries, and we don't call ConfigureAwait(false) elsewhere, because we don't create a synchonization context and neither does asp.net core. 
You could certainly make the case that we should. But then we should do it everywhere and turn on an analyzer rule to enforce it. I'm not sure the added noise would be worth it.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 6 above)
Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:81]: Start _should_ be called first. This is just contract enforcement. The class cannot be used until Started, which makes it special, and it will tell you if try to use it before Started (i.e. in a unit test). (Or if ever we did publish this a general-purpose library, though I'm contradicting my comment on ConfigureAwait :)) Not a big deal either way.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 7 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:198]: Okay, makes sense. Thanks for the background!

Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs]: Removed in 1079d38.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 5 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: Good point. Removed in 4e00f98 and c1c9e77.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs (hunk 6 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs:85]: You're right - I found this useful when debugging unit tests. Added the check back in in c583087.
